web_src/js/features/codeeditor.js

```diff
@@ -105,14 +105,26 @@ export async function createMonaco(textarea, filename, editorOpts) {
   monaco.languages.register({id: 'vs.editor.nullLanguage'});
   monaco.languages.setLanguageConfiguration('vs.editor.nullLanguage', {});
 
+  // TODO: there must be a better way to preserve CRLF in the template rendering
```

Comment:
I guess there is no other way besides using JS.

HTML standard does play tricks with CRLF/LF for textarea value.

Reply:
I think it's a self-caused issue by gitea's HTML rendering. But no idea where it happens.

Reply:
* https://github.com/go-gitea/gitea/pull/24745#issuecomment-1548917162

> For historical reasons, the element's value is normalized in three different ways for three different purposes.
> The raw value is the value as it was originally set. It is not normalized. 
> The API value ... It is normalized so that line breaks use U+000A LINE FEED (LF) characters. 

`textarea.value` is the API value IIRC.

Reply:
I also tried as `data-value` and the `\r` was stripped as well there. So it's not related to `textarea.value`.

Reply:
I do not see so. I think everything just works well.

![image](https://github.com/go-gitea/gitea/assets/2114189/1b3bd6d1-5c90-4747-8c70-2b59797c8dc4)

Reply:
Try with a `ctx["Data"]` value as well.

Reply:
I do not see anything wrong. Technically speaking there shouldn't be any surprise. Maybe you misunderstand something?

![image](https://github.com/go-gitea/gitea/assets/2114189/5c0d9abf-873a-455f-8a14-3203d4fb2ac9)

Reply:
Your result is expected, strange. When I tried it, it was outputting literal `\r` and `\n` into the HTML. I wonder why is it escaping it in your case?

Reply:
Golang template does some URL-encoding for some attributes (like `href`, `data`) automatically.

If you would like to emit URL-encoded value for some `data-xxx` attributes, it needs `{{.Value | QueryEscape}}`

Reply:
On the other hand, `\r` needs no escaping in HTML, only the 5 known entities need it, e.g. [`html.EscapeString`](https://pkg.go.dev/html#EscapeString), so it would escape unnecessarily.

Reply:
@wxiaoguang apparently go escapes `data` attribute differently then `data-foo` or textarea content, you will be able to reproduce on current branch which has:

````html
<textarea data="{{.FileContent}}" data-foo="{{.FileContent}}">{{.FileContent}}</textarea>
```

The code logs all three values. With [this file](https://github.com/silverwind/symlink-test/blob/master/eol-crlf.js) for me it logs:

```
"a = 1\nb = 2\n"
"a%20=%201%0d%0ab%20=%202%0d%0a" 
"a = 1\nb = 2\n"
```

Reply:
https://go.dev/play/p/vAw0hXzPXMZ also shows that `data` vs. `data-foo` are escaped differently.

Reply:
I also added a devtest showcase in https://github.com/go-gitea/gitea/pull/27141/commits/d6fd3f6cfeabfaa0b3fd1a724dbbb47e98f694d5. if you go to gitea-ui, it will log the incorrect `"a\nb"` into the console.

Reply:
see https://github.com/go-gitea/gitea/pull/27141#discussion_r1331568013

Reply:
1. QueryEscape is to escape query parameters in a URL, it's the wrong kind of escape in a HTML attribute. golang already does this correctly implicitely.
2. I fail to see how that is related to the problem. I have shown that regular go template does not have the problem, but ours does.

Maybe I'll pass the value through `pageData` if I can't find the source of the problem. I do have a feeling the regression may have come in https://github.com/go-gitea/gitea/commit/722dab5286a001a0bd367d42258a8613a944e1c1, but it's really hard to debug.

Reply:
The standard has clarified that there is no CR but only LF when parsing HTML. I do not know how to explain more.

Nothing is unexpected, everything is right from HTML/Golang/GoTemplate/Gitea side.

The only wrong thing is that you shouldn't emit "\r" into HTML directly if you want to keep it.

Reply:
> 2\. I fail to see how that is related to the problem. I have shown that regular go template does not have the problem, but ours rendering does, so it's a bug in our rendering.

Your code is wrong. I have already shown that our code renders "\r\n" correctly: https://github.com/go-gitea/gitea/pull/27141#discussion_r1331320996

Reply:
Ah, you mean the browser discards it during HTML parsing? I will check with curl to confirm then. If this is true, then yes, we need to encode it to preserve the `\r`.

Reply:
curl confirms, it's output correctly. Damn browsers and their silly parsers:

```
curl -s http://localhost:3500/devtest/gitea-ui | grep data-foo= | xxd
00000000: 3c64 6976 2063 6c61 7373 3d22 7061 6765  <div class="page
00000010: 2d63 6f6e 7465 6e74 2064 6576 7465 7374  -content devtest
00000020: 2075 6920 636f 6e74 6169 6e65 7222 2064   ui container" d
00000030: 6174 612d 666f 6f3d 2261 0d0a            ata-foo="a..

Reply:
Fixed with https://github.com/go-gitea/gitea/pull/27141/commits/f394554ba9aa862c7cbd27e6592a46d0476b2309.